File: cpp/src/runtime/execution.cpp
#include "maglev/runtime/execution.h"

#include <filesystem>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <utility>

#include "maglev/runtime/util.h"

namespace maglev::execution {

namespace {

void ensure_success(const CommandResult& result, const std::string& action) {
    if (result.exit_code != 0) {
        std::ostringstream stream;
        stream << action << " failed";
        if (!trim(result.stdout_text).empty()) {
            stream << ": " << trim(result.stdout_text);
        } else if (!trim(result.stderr_text).empty()) {
            stream << ": " << trim(result.stderr_text);
        }
        throw std::runtime_error(stream.str());
    }
}

}  // namespace

bool ask_for_approval(const std::string& question, bool auto_approve) {
    if (auto_approve) {
        return true;
    }

    std::cout << question << " [y/N] " << std::flush;
    std::string answer;
    std::getline(std::cin, answer);
    const auto normalized = to_lower(trim(answer));
    return normalized == "y" || normalized == "yes";
}

WorkspaceExecutor::WorkspaceExecutor(std::filesystem::path workspace_root) : workspace_root_(std::move(workspace_root)) {}

void WorkspaceExecutor::apply_edits(const std::vector<EditProposal>& edits) const {
    for (const auto& edit : edits) {
        std::string error;
        if (!write_text_file(workspace_root_ / edit.path, edit.content, error)) {
            throw std::runtime_error(error);
        }
    }
}

void WorkspaceExecutor::run_checks() const {
    ensure_success(run_script("npm run check", workspace_root_), "local checks");
}

void WorkspaceExecutor::git_commit(const CommitMessageProposal& proposal) const {
    ensure_success(run_script("git add .", workspace_root_), "git add");

    std::ostringstream script;
    script << "git commit -m " << quote_shell_argument(proposal.title);
    if (proposal.body && !proposal.body->empty()) {
        script << " -m " << quote_shell_argument(*proposal.body);
    }
    ensure_success(run_script(script.str(), workspace_root_), "git commit");
}

void WorkspaceExecutor::git_push() const {
    const auto branch = current_branch();
    if (branch.empty()) {
        throw std::runtime_error("failed to resolve current branch");
    }

    ensure_success(run_script("git push origin " + quote_shell_argument(branch), workspace_root_), "git push");
}

std::string WorkspaceExecutor::git_status_short() const {
    auto result = run_script("git status --short", workspace_root_);
    ensure_success(result, "git status --short");
    const auto output = trim(result.stdout_text);
    return output.empty() ? std::string("No uncommitted changes.") : output;
}

std::string WorkspaceExecutor::current_branch() const {
    auto result = run_script("git branch --show-current", workspace_root_);
    ensure_success(result, "git branch --show-current");
    return trim(result.stdout_text);
}

std::string WorkspaceExecutor::head_commit() const {
    auto result = run_script("git rev-parse HEAD", workspace_root_);
    ensure_success(result, "git rev-parse HEAD");
    return trim(result.stdout_text);
}

std::vector<std::string> WorkspaceExecutor::changed_files() const {
    auto result = run_script("git status --short", workspace_root_);
    ensure_success(result, "git status --short");

    std::vector<std::string> files;
    for (const auto& line : split_lines(result.stdout_text)) {
        const auto trimmed_line = trim(line);
        if (trimmed_line.size() <= 3) {
            continue;
        }
        files.push_back(trimmed_line.substr(3));
    }
    return files;
}

std::string DeployExecutor::build_remote_command(const DeployRequestProposal& request) {
    return request.restart_command ? "cd " + request.repo_path + " && git pull origin " + request.branch + " && " +
                                         *request.restart_command
                                   : "cd " + request.repo_path + " && git pull origin " + request.branch;
}

std::string DeployExecutor::preview(const DeployRequestProposal& request) {
    const auto remote_command = build_remote_command(request);
    return "Host: " + request.host + "\nSSH command: ssh " + request.host + "\nRemote command: " + remote_command;
}

DeployResult DeployExecutor::execute(const DeployRequestProposal& request) {
    const auto remote_command = build_remote_command(request);
    std::ostringstream script;
    script << "ssh " << quote_shell_argument(request.host) << " " << quote_shell_argument(remote_command);
    auto result = run_script(script.str(), std::filesystem::current_path());

    DeployResult deploy_result;
    deploy_result.success = result.exit_code == 0;
    deploy_result.host = request.host;
    deploy_result.branch = request.branch;
    deploy_result.health = result.exit_code == 0 ? "ok" : "failed";
    deploy_result.logs.push_back(result.stdout_text);
    deploy_result.logs.push_back(result.stderr_text);
    return deploy_result;
}

}  // namespace maglev::execution

File: src/execution.rs
use std::fs;
use std::io::{self, Write};
use std::path::Path;
use std::process::Command;

use anyhow::{bail, Context, Result};

use crate::contracts::{CommitMessageProposal, DeployRequestProposal, DeployResult, EditProposal};

pub fn ask_for_approval(question: &str, auto_approve: bool) -> Result<bool> {
    if auto_approve {
        return Ok(true);
    }

    print!("{question} [y/N] ");
    io::stdout().flush().context("failed to flush stdout")?;

    let mut answer = String::new();
    io::stdin().read_line(&mut answer).context("failed to read approval input")?;
    Ok(matches!(answer.trim().to_lowercase().as_str(), "y" | "yes"))
}

pub fn apply_edits(workspace_root: &str, edits: &[EditProposal]) -> Result<()> {
    for edit in edits {
      let file_path = Path::new(workspace_root).join(&edit.path);
      fs::write(&file_path, &edit.content).with_context(|| format!("failed to write {}", file_path.display()))?;
    }

    Ok(())
}

pub fn run_checks(workspace_root: &str) -> Result<()> {
    let status = Command::new("npm")
        .args(["run", "check"])
        .current_dir(workspace_root)
        .status()
        .context("failed to spawn npm run check")?;

    if !status.success() {
        bail!("local checks failed");
    }

    Ok(())
}

pub fn git_commit(workspace_root: &str, proposal: &CommitMessageProposal) -> Result<()> {
    let add_status = Command::new("git")
        .args(["add", "."])
        .current_dir(workspace_root)
        .status()
        .context("failed to spawn git add")?;
    if !add_status.success() {
        bail!("git add failed");
    }

    let mut commit = Command::new("git");
    commit.args(["commit", "-m", &proposal.title]).current_dir(workspace_root);
    if let Some(body) = &proposal.body {
        commit.args(["-m", body]);
    }

    let commit_status = commit.status().context("failed to spawn git commit")?;
    if !commit_status.success() {
        bail!("git commit failed");
    }

    Ok(())
}

pub fn git_push(workspace_root: &str) -> Result<()> {
    let branch_output = Command::new("git")
        .args(["branch", "--show-current"])
        .current_dir(workspace_root)
        .output()
        .context("failed to read current branch")?;
    if !branch_output.status.success() {
        bail!("failed to resolve current branch");
    }

    let branch = String::from_utf8_lossy(&branch_output.stdout).trim().to_string();
    let push_status = Command::new("git")
        .args(["push", "origin", &branch])
        .current_dir(workspace_root)
        .status()
        .context("failed to spawn git push")?;

    if !push_status.success() {
        bail!("git push failed");
    }

    Ok(())
}

pub fn preview_deploy(request: &DeployRequestProposal) -> String {
    let remote_command = match &request.restart_command {
        Some(restart) => format!("cd {} && git pull origin {} && {}", request.repo_path, request.branch, restart),
        None => format!("cd {} && git pull origin {}", request.repo_path, request.branch),
    };

    format!(
        "Host: {}\nSSH command: ssh {}\nRemote command: {}",
        request.host, request.host, remote_command
    )
}

pub fn deploy(request: &DeployRequestProposal) -> Result<DeployResult> {
    let remote_command = match &request.restart_command {
        Some(restart) => format!("cd {} && git pull origin {} && {}", request.repo_path, request.branch, restart),
        None => format!("cd {} && git pull origin {}", request.repo_path, request.branch),
    };

    let output = Command::new("ssh")
        .arg(&request.host)
        .arg(&remote_command)
        .output()
        .context("failed to spawn ssh")?;

    Ok(DeployResult {
        success: output.status.success(),
        host: request.host.clone(),
        branch: request.branch.clone(),
        health: if output.status.success() { "ok".into() } else { "failed".into() },
        logs: vec![
            String::from_utf8_lossy(&output.stdout).to_string(),
            String::from_utf8_lossy(&output.stderr).to_string(),
        ],
    })
}
